README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# PropluviaR

<!-- badges: start -->

[![R-CMD-check](https://github.com/richaben/PropluviaR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/richaben/PropluviaR/actions/workflows/R-CMD-check.yaml)
[![License: GPL
v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)
<!-- badges: end -->

Le package `PropluviaR` vise à faciliter la récupération des données
[Propluvia](https://propluvia.developpement-durable.gouv.fr/propluviapublic/accueil)
sur les restrictions d’eau à partir du site du gouvernement.

Il utilise l’API du site pour récupérer notamment les zones d’alerte
pour les eaux superficielles et souterraines à l’échelle de la France à
une date spécifiée.

L’export des données collectées est possible par la création d’une
couche `.gpkg` pour une utilisation SIG.

## Installation ⚙️

Pour l’installer :

``` r
if (!require("devtools")) install.packages("devtools")
devtools::install_github("richaben/PropluviaR")
```

Pour le charger :

``` r
library(PropluviaR)
```

## Exemples

### Téléchargements des données 🖥️

-   Récupération des données du jour, sans export :

``` r
propluvia_data_jour <- 
  get_propluvia(export = FALSE)
```

-   Récupération des données à une date précise (ici le 23 juillet
    2022), sans export :

``` r
propluvia_data_20220723 <- 
  get_propluvia(date = '2022-07-23', export = FALSE)
```

-   Export au format geopackage `.gpkg` (à récupérer dans le dossier de
    travail) :

``` r
get_propluvia(export = TRUE)
```

### Un exemple de carte 🗺️

``` r
#devtools::install_github("MaelTheuliere/COGiter")
library(COGiter)

# shape pour les départements FR métropolitaine
dept_shp <- 
  COGiter::departements_geo %>% 
  dplyr::filter(!DEP %in% c("971", "972", "973", "974", "976"))

# les données Propluvia du jour !
propluvia_data_jour <- 
  get_propluvia(export = FALSE)

# palette de couleur
pal_colZone <- 
  setNames(unique(propluvia_data_jour$colSeuil),
           unique(propluvia_data_jour$nomNiveau))

# le graphique
plot_propluvia <-
  ggplot2::ggplot() +
  ggplot2::geom_sf(data = (propluvia_data_jour %>% 
                             dplyr::filter(typeZone == 'SUP')),
                   ggplot2::aes(fill=nomNiveau),
                   col='black',size=0.1) +
  ggplot2::scale_fill_manual(values = pal_colZone, "Zones d'alerte") +
  ggplot2::geom_sf(data = dept_shp, col='black', size=0.6, fill=NA) +
  ggplot2::labs(title = paste0("Carte Propluvia au ", format(Sys.time(),"%d/%m/%Y"), " pour les eaux superficielles."), 
                caption = "Source: http://propluvia.developpement-durable.gouv.fr") +
  ggplot2::theme_void() +
  ggplot2::theme(title = ggplot2::element_text(face = 'bold'),
                 legend.position = "bottom")

plot_propluvia
```

<img src="man/figures/PropluviaR_map_exemple.png" width="100%" />


NEWS.md
# PropluviaR 0.1.0

* Created the package
* Added the `get_propluvia()` function
* Added the `get_propluvia_dpt()` function
* Added a vignette
* Added a `NEWS.md` file to track changes to the package.


R/get_propluvia.R
#' Télécharge l'ensemble des données Propluvia à date du jour ou spécifiée
#'
#' @encoding UTF-8
#' @description [get_propluvia()] permet de télécharger les données Propluvia pour la France à une date donnée (par défaut la date du jour).
#' Les données contiennent la géométrie des zones d'alertes.
#' L'argument `export` de la fonction permet de sauvegarder les données au format [`.gpkg`] pour utilisation SIG.
#' Les données mises en forme contiennent une colonne avec le code couleur des zones d'alerte.
#'
#' @param date Date. Par défaut, la date du jour est utilisée. Si elle est renseignée, doit être au format ["\%Y-\%m-\%d"] (ex: 2022-07-23).
#' @param export Valeur logique (par défaut `FALSE`). Si `TRUE`, créer une couche au format [`.gpkg`] dans le répertoire de travail.
#'
#' @return data.frame au format `sf`. Contient les geometry des zones d'alerte.
#' @export
#'
#' @importFrom cli cli cli_h1 cli_abort cli_alert_info cli_alert_danger
#' @importFrom curl has_internet
#' @importFrom dplyr bind_cols mutate case_when
#' @importFrom geojsonsf geojson_sf
#' @importFrom glue glue
#' @importFrom httr GET http_error message_for_status content
#' @importFrom jsonlite fromJSON toJSON
#' @importFrom sf st_as_sf write_sf
#'
#' @examples
#' \dontrun{
#' data_jour <- get_propluvia(export = FALSE)
#' }

get_propluvia <- function(date, export = FALSE){
  
  cli::cli(
    {cli::cli_h1("R\u00e9cup\u00e9ration des donn\u00e9es Propluvia")}
  )
  
  # Check internet connexion
  if (!curl::has_internet()) {
    message("Pas de connexion internet.")
    return(FALSE)
  }
  
  date <- if(missing(date)){format(Sys.time(),'%Y-%m-%d')} else {date}
  
  if(is.na(as.Date(date, format = '%Y-%m-%d'))){
    stop(cli::cli_abort("Mauvais format de date ! (ex: 2022-07-23)"))
  }
  
  # api_url_propluvia
  api_url_propluvia <- glue::glue('https://eau.api.agriculture.gouv.fr/apis/propluvia/zones/{date}')
  
  cli::cli_alert_info("T\u00e9l\u00e9chargement {.url {api_url_propluvia}}")
  
  response <-
    httr::GET(url = api_url_propluvia)
  
  if(httr::http_error(response)) {
    httr::message_for_status(response)
    cli::cli_alert_danger("V\u00e9rifier l\'url dans le navigateur ({.url {api_url_propluvia}})")
  }
  
  cli::cli_alert_info("Assemblage des donn\u00e9es")
  
  response <-
    response %>%
    httr::content('text') %>%
    jsonlite::fromJSON()
  
  data_geometry <-
    response$contourZone$geometry %>%
    jsonlite::toJSON() %>%
    geojsonsf::geojson_sf() %>%
    dplyr::bind_cols(numeroNiveau = as.factor(response$numeroNiveau)) %>%
    dplyr::bind_cols(typeZone = response$typeZone) %>%
    dplyr::bind_cols(nomZone = response$nomZone) %>%
    dplyr::mutate(nomNiveau = dplyr::case_when(numeroNiveau == "1" ~ 'Vigilance',
                                               numeroNiveau == "3" ~ 'Alerte',
                                               numeroNiveau == "4" ~ 'Alerte renforc\u00e9e',
                                               numeroNiveau == "5" ~ 'Crise'),
                  nomNiveau = factor(nomNiveau, levels = c("Vigilance",
                                                           "Alerte",
                                                           "Alerte renforc\u00e9e",
                                                           "Crise"),
                                     ordered = T),
                  colSeuil = dplyr::case_when(numeroNiveau == "1" ~ '#EEEEEE',
                                              numeroNiveau == "3" ~ '#FAE805',
                                              numeroNiveau == "4" ~ '#FF9900',
                                              numeroNiveau == "5" ~ '#FF0000')) %>%
    sf::st_as_sf()
  
  # Export in geopackage
  if(!export == F) {
    cli::cli_alert_info("Export en gpkg {.url {getwd()}}")
    sf::write_sf(obj = data_geometry, dsn = glue::glue("propluvia_{date}.gpkg", append = TRUE))
  }
  
  data_geometry
  
}



R/get_propluvia_dpt.R
#' Télécharge les données Propluvia pour un département et à date du jour (ou spécifiée)
#'
#' @encoding UTF-8
#' @description [get_propluvia_dpt()] permet de télécharger les données Propluvia pour un département à une date donnée (par défaut la date du jour).
#' Les données contiennent la géométrie des zones d'alertes, et les infos sur les arrêtés en vigueur dans le département.
#' L'argument `export` de la fonction permet de sauvegarder les données au format [`.gpkg`] pour utilisation SIG.
#' Les données mises en forme contiennent également une colonne avec le code couleur des zones d'alerte.
#'
#' @param dpt Numérique. Correspond au numéro de département, par ex. 14 pour le Calvados.
#' @param date Date. Par défaut, la date du jour est utilisée. Si elle est renseignée, doit être au format ["\%Y-\%m-\%d"] (ex: 2022-07-23).
#' @param export Valeur logique (par défaut `FALSE`). Si `TRUE`, créer une couche au format [`.gpkg`] dans le répertoire de travail.
#'
#' @return data.frame au format `sf`. Contient les informations relatives aux arrêtés en vigueur dans le département (numéro, date de signature, date de validité).
#'     Les niveaux de restriction avec leurs dénominations (Vigilance, Alerte, Alerte renforcée, Crise) et le code couleur associé.
#'     Les noms des zones d'alertes.
#'     Les types de zones : superficielles (`SUP`) et/ou souterraines (`SOU`).
#'     Les geometries des zones d'alerte (multipolygones au format WGS84, crs 4326).
#' @export
#'
#' @importFrom cli cli cli_h1 cli_abort cli_alert_info cli_alert_danger
#' @importFrom curl has_internet
#' @importFrom dplyr select mutate left_join bind_cols case_when
#' @importFrom geojsonsf geojson_sf
#' @importFrom glue glue
#' @importFrom httr GET http_error message_for_status content
#' @importFrom jsonlite fromJSON toJSON
#' @importFrom sf st_as_sf write_sf
#'
#' @examples
#' \dontrun{
#' data_jour_dpt14 <- get_propluvia_dpt(dpt = 27,export = FALSE)
#' }

get_propluvia_dpt <- function(dpt, date, export = FALSE){
  
  cli::cli(
    {cli::cli_h1("R\u00e9cup\u00e9ration des donn\u00e9es Propluvia du d\u00e9partement")}
  )
  
  # Check internet connexion
  if (!curl::has_internet()) {
    message("Pas de connexion internet.")
    return(FALSE)
  }
  
  date <- if(missing(date)){format(Sys.time(),'%Y-%m-%d')} else {date}
  
  if(is.na(as.Date(date, format = '%Y-%m-%d'))){
    stop(cli::cli_abort("Mauvais format de date ! (ex: 2022-07-23)"))
  }
  
  if(missing(dpt)) {
    stop(cli::cli_abort("Num\u00e9ro de d\u00e9partement manquant !"))
  }
  
  # api_url_propluvia
  api_url_propluvia <- glue::glue('https://eau.api.agriculture.gouv.fr/apis/propluvia/arretes/{date}/departement/{dpt}')
  
  cli::cli_alert_info("T\u00e9l\u00e9chargement {.url {api_url_propluvia}}")
  
  response <-
    httr::GET(url = api_url_propluvia)
  
  if(httr::http_error(response)) {
    httr::message_for_status(response)
    cli::cli_alert_danger("V\u00e9rifier l\'url dans le navigateur ({.url {api_url_propluvia}})")
  }
  
  cli::cli_alert_info("Assemblage des donn\u00e9es")
  
  response <-
    response %>%
    httr::content('text') %>%
    jsonlite::fromJSON()
  
  part_date_arrete <-
    response %>%
    as.data.frame() %>%
    dplyr::select(-restrictions)
  
  tab_zoneAlerte <-
    response$restrictions[[1]]$zoneAlerte %>%
    dplyr::select(-communes,-contourZone,-departements) %>%
    dplyr::mutate(codeInseeDepartement = as.character(rep({dpt},nrow(.)))) %>%
    dplyr::left_join(part_date_arrete) %>%
    dplyr::bind_cols(niveauRestriction = response$restrictions[[1]]$niveauRestriction) %>%
    dplyr::bind_cols(nomNiveau = response$restrictions[[1]]$nomNiveau) %>%
    dplyr::bind_cols(response$restrictions[[1]]$zoneAlerte$contourZone$geometry %>%
                       jsonlite::toJSON() %>%
                       geojsonsf::geojson_sf()) %>%
    dplyr::mutate(nomNiveau = factor(nomNiveau, levels = c("Vigilance",
                                                           "Alerte",
                                                           "Alerte renforc\u00e9e",
                                                           "Crise"),
                                     ordered = T),
                  colSeuil = dplyr::case_when(niveauRestriction  == "1" ~ '#EEEEEE',
                                              niveauRestriction  == "3" ~ '#FAE805',
                                              niveauRestriction  == "4" ~ '#FF9900',
                                              niveauRestriction  == "5" ~ '#FF0000')) %>%
    sf::st_as_sf()
  
  # Export in geopackage
  if(!export == F) {
    cli::cli_alert_info("Export en gpkg {.url {getwd()}}")
    sf::write_sf(obj = tab_zoneAlerte, dsn = glue::glue("propluvia_dpt{dpt}_{date}.gpkg", append = TRUE))
  }
  
  tab_zoneAlerte
  
}



